site/_posts/2023-07-11-get-more-value-out-of-fusionauth.md

```diff
@@ -0,0 +1,321 @@
+
+<br>
+
+## Custom Validation with FusionAuth APIs
+*“With great power, comes great responsibility!”*
+
+One of FusionAuth’s strengths is its flexibility. However, with the flexibility comes some complexity. To create a user in the FusionAuth front-end, you only need one piece of information. You will need to supply either the username or the email. (If you are using multiple tenants, you will need the tenant ID as well.) The front end gives you other optional fields to fill in as well, however none are required. When using the [Create a User API](/docs/v1/tech/apis/users#create-a-user), only the username or email is still required but there are also over 30 additional request parameters available for customization. Creating a custom front end around the APIs will not only allow you to choose which of the username or email fields to use, but also what other meta-data is appropriate and required for your application. Below is sample code for creating a user using the FusionAuth API.
+
+<br>
+
+### **Setup**
+First, we need to create a few objects to hold some values. By constructing the classes and decorating properties with the proper attributes, we can still use the strongly typed objects and standard naming conventions in our C# application while also using the JSON Serializer to easily format the request body when it is time.
+
+```cs
+//For User object
+public class User
+{
+    [JsonProperty(PropertyName = "user")]
+    public UserProperties UserProperties;
+}
+
+public class UserProperties
+{
+    [JsonProperty(PropertyName = "firstname")]
+    public string FirstName;
+    [JsonProperty(PropertyName = "lastname")]
+    public string LastName;
+    [JsonProperty(PropertyName = "email")]
+    public string Email;
+    [JsonProperty(PropertyName = "password")]
+    public string Password;
+}
+```
+
+<br>
+
+### **Input**
+Next, we need to take the input from the user and assign them to the values in our User object. In this case, I am performing those actions on a click event from a windows form. We can also take this chance to validate any information deemed relevant to your application. The logic in the IsUserValid function can be edited to suit your needs. Once the validation is successful, we will then pass the user object to the method that will submit our request to the FusionAuth server.
+
+```cs
+private async void btnCreateUser_Click(object sender, EventArgs e)
+{
+    //FusionAuthClient is a class that contains the logic to call the FusionAuth APIs
+    var faClient = new FusionAuthClient();
+
+    UserProperties userProperties = new UserProperties()
+    {
+        Email = txtEmail.Text,
+        FirstName = txtFirstName.Text,
+        LastName = txtLastName.Text,
+        Password = txtPassword.Text
+        //add additional properties here
+    };
+
+    User userToCreate = new User()
+    {
+        UserProperties = userProperties
+    };
+
+    //perform custom validation logic in the IsUserValid method
```

Comment:
Any reason not to share the `IsUserValid` logic? Seems like that would be interesting to see.